janitor/functions.py

```diff
@@ -4332,56 +4350,51 @@ def fill_direction(
-        can be either `down`(default), `up`, `updown`(fill up then down) and
+        can be either `down` (default), `up`, `updown` (fill up then down) and
         `downup` (fill down then up).
     :param limit: number of consecutive null values to forward/backward fill.
         Value must `None` or greater than 0.
     :returns: A pandas dataframe with modified column(s).
-    :raises ValueError: if ``directions`` dictionary is empty.
     :raises ValueError: if column supplied is not in the dataframe.
-    :raises ValueError: if direction supplied is not one of `down`,`up`,
+    :raises ValueError: if direction supplied is not one of `down`, `up`,
         `updown`, or `downup`.
+
+    .. # noqa: DAR402
     """
     df = df.copy()
-    # check that dictionary is not empty
     if not directions:
-        raise ValueError("A mapping of columns with directions is required.")
-
-    # check that the right columns are provided
-    # should be removed once the minimum Pandas version is 1.1,
-    # as Pandas loc will raise a KeyError if columns provided do not exist
-    wrong_columns_provided = set(directions).difference(df.columns)
-    if any(wrong_columns_provided):
-        if len(wrong_columns_provided) > 1:
-            outcome = ", ".join(f"'{word}'" for word in wrong_columns_provided)
-            raise ValueError(
-                f"Columns {outcome} do not exist in the dataframe."
-            )
-        outcome = "".join(wrong_columns_provided)
-        raise ValueError(f"Column {outcome} does not exist in the dataframe.")
+        return df
 
-    # check that the right directions are provided
-    set_directions = {"up", "down", "updown", "downup"}
+    check("directions", directions, [dict])
 
-    # linter throws an error when I use dictionary.values()
-    # it assumes that dictionary is a dataframe
-    directions_values = [value for key, value in directions.items()]
-    wrong_directions_provided = set(directions_values).difference(
-        set_directions
-    )
-    if any(wrong_directions_provided):
-        raise ValueError(
-            """The direction should be a string and should be one of `up`,
-            `down`, `updown`, or `downup`."""
-        )
+    if limit is not None:
+        check("limit", limit, [int])
+        # pandas raises error if limit is not greater than zero
+        # so no need for a check on pyjanitor's end
+
+    check_column(df, directions)
+
+    for _, direction in directions.items():
+        if direction not in {"up", "down", "updown", "downup"}:
+            raise ValueError(
+                """
+                The direction should be a string and should be one of
+                `up`, `down`, `updown`, or `downup`.
+                """
+            )
 
+    # TODO: option to specify limit per column; current implementation
+    # is one `limit` for all the columns. Might need refactoring, or an
+    # API change.
```

Comment:
This is a classic "do we apply this function to one column, or multiple columns" situation! 

As the library matures, I've grown to realize that having singular and plural functions is not necessarily the best, so allowing a function to be applied to multiple columns simultaneously is probably the better way. There must be some way we can "compose" them together, I think, such that we get the desired behaviour. I'm kind of inspired by what I see in JAX, where we have `vmap` to map a function over an array axis. This makes composing functions together more natural.

Maybe a better solution to this problem is to provide a `.map_columns()` function, which allows us to map any pyjanitor function across columns. Let's talk about this in another setting. For now, I like what you've done here, @samukweku, in that you've left this API piece as a TODO to think about later on.